File: probs/15686.py
import sys
input = sys.stdin.readline

input = open('input.txt', 'r').readline

from itertools import combinations as c

N, M = map(int, input().split())
houses = []
chickens = []
city = [list(map(int, input().split())) for _ in range(N)]
for i, j in [(i, j) for i in range(N) for j in range(N)]:
    if city[i][j] == 1:
        houses.append((i, j))
    elif city[i][j] == 2:
        chickens.append((i, j))

distances = []
for ci, cj in chickens:
    distance = []
    for hi, hj in houses:
        distance.append(abs(ci-hi) + abs(cj-hj))
    distances.append(distance)

ans = 10**8
for dists in c(distances, M):
    ch_dist = [10**8] * len(houses)
    for dist in dists:
        for i, d in enumerate(dist):
            ch_dist[i] = min(ch_dist[i], d)
    ans = min(ans, sum(ch_dist))
print(ans)

File: probs/15686.cpp
#include <bits/stdc++.h>

using namespace std;
using pi = pair<int, int>;

const int INF = 1e9;

int N, M, ans = INF;
vector<pi> H, C;
vector<vector<int>> D;

int dist(pi x, pi y)
{
    return abs(x.first - y.first) + abs(x.second - y.second);
}

void bt(int i, int n, vector<int> cdist)
{
    if (n == M) {
        int s = 0;
        for (int d : cdist) s += d;
        ans = min(ans, s);
        return;
    }

    if (i == C.size()) return;

    bt(i+1, n, cdist);
    for (int j = 0; j < H.size(); j++)
        cdist[j] = min(cdist[j], D[i][j]);
    bt(i+1, n+1, cdist);
}

int main()
{
    ios_base::sync_with_stdio(0);
    cin.tie(0);

    freopen("input.txt", "r", stdin);

    cin >> N >> M;
    H.reserve(2*N); C.reserve(13);
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            int ipt; cin >> ipt;
            if (ipt == 1) H.emplace_back(i, j);
            else if (ipt == 2) C.emplace_back(i, j);
        }
    }

    D.resize(C.size());
    for (int i = 0; i < C.size(); i++) {
        D[i].resize(H.size());
        for (int j = 0; j < H.size(); j++) {
            D[i][j] = dist(C[i], H[j]);
        }
    }

    bt(0, 0, vector<int>(H.size(), INF));
    cout << ans << '\n';

    return 0;
}
